### PR TITLE
backend update

### PR DESCRIPTION
**Pull request by ddjnw1yu**
- create new APIs for getting, passing, filtering and downloading gen3 metadata.
- test created APIs

### CODE REVIEW

File: app/main.py [hunk 1]
```diff
@@ -76,3 +102,131 @@ def get_share_link():
...
+
+@app.route("/dictionary", methods=['GET', 'POST'])
+# Get all dictionary node from Gen3 Data Commons
+def dictionary():
+    res = requests.get(
+        f'{Gen3Config.GEN3_ENDPOINT_URL}/api/v0/submission/_dictionary', headers=HEADER)
+
+    json_data = json.loads(res.content)
+    dictionary_list = []
+    for ele in json_data['links'][2:30]:
```
Comment (alan-wu) [inline app/main.py]: Is there a reason to include items 2-30 only?

File: app/main.py [hunk 2]
```diff
@@ -76,3 +102,131 @@ def get_share_link():
...
+
+@app.route('/nodes/<node_type>', methods=['GET', 'POST'])
+# Exports all records in a dictionary node
+def export_node(node_type):
+    post_data = request.get_json()
+    program = post_data.get('program')
+    project = post_data.get('project')
+    format = post_data.get('format')
+    res = requests.get(
+        f'{Gen3Config.GEN3_ENDPOINT_URL}/api/v0/submission/{program}/{project}/export/?node_label={node_type}&format={format}', headers=HEADER)
```
Comment (alan-wu) [inline app/main.py]: Would this work with GET method with the program, project and format being undefined?

File: app/main.py [hunk 3]
```diff
@@ -76,3 +102,131 @@ def get_share_link():
...
+@app.route('/records/<uuid>', methods=['GET', 'POST'])
+# Exports one or more records, use comma to separate the uuids
+# e.g. uuid1,uuid2,uuid3
+def export_record(uuid):
+    post_data = request.get_json()
+    program = post_data.get('program')
+    project = post_data.get('project')
+    format = post_data.get('format')
+    res = requests.get(
+        f'{Gen3Config.GEN3_ENDPOINT_URL}/api/v0/submission/{program}/{project}/export/?ids={uuid}&format={format}', headers=HEADER)
```
Comment (alan-wu) [inline app/main.py]: Would this work with GET method with the program, project and format being undefined?

File: app/main.py [hunk 4]
```diff
@@ -76,3 +102,131 @@ def get_share_link():
...
+    return new_json_data
+
+
+@app.route("/<program>/project", methods=['GET'])
+# Get all projects information from Gen3 Data Commons
+def project(program):
+    res = requests.get(
+        f'{Gen3Config.GEN3_ENDPOINT_URL}/api/v0/submission/{program}', headers=HEADER)
+
+    json_data = json.loads(res.content)
```
Comment (alan-wu) [inline app/main.py]: Should handle the case when the requests.get fails, in which case the res.content will not be valid and will cause uncaught exception in the backend.

File: app/main.py [hunk 5]
```diff
@@ -76,3 +102,131 @@ def get_share_link():
...
+    new_json_data = {'project': project_list}
+    return new_json_data
+
+
+@app.route("/dictionary", methods=['GET', 'POST'])
+# Get all dictionary node from Gen3 Data Commons
+def dictionary():
+    res = requests.get(
+        f'{Gen3Config.GEN3_ENDPOINT_URL}/api/v0/submission/_dictionary', headers=HEADER)
+
```
Comment (alan-wu) [inline app/main.py]: Should handle the case when the requests.get fails, in which case the res.content will not be valid and will cause uncaught exception in the backend.

**Review by alan-wu — CHANGES_REQUESTED** [review]
It is a good start, I have asked some question and made a few requests.

File: app/main.py (hunk 1 above)
Comment (ddjnw1yu) [reply app/main.py]: The items that not included are "root", "data_release", "_all". They are not dictionary nodes, and I want to see the nodes only when I test the code. It has no meaning. I just forgot to remove it after I finished.

File: app/main.py (hunk 2 above)
Comment (ddjnw1yu) [reply app/main.py]: Yes, that will be a problem if the program, project and format are undefined. I have modified the error handling code based on different situations.

File: app/main.py (hunk 4 above)
Comment (ddjnw1yu) [reply app/main.py]: Did not consider this situation at first. Has added try except to those APIs.

File: app/main.py [hunk 6]
```diff
@@ -76,3 +105,181 @@ def get_share_link():
...
+    if json_data["data"] == None:
+        abort(NOT_FOUND, description=json_data["errors"])
+    else:
+        if json_data["data"][f"{node_type}"] == []:
+            abort(NOT_FOUND)
+        else:
+            return res.content
+
+
+@app.route("/<program>/<project>/<uuid>/<format>/download", methods=["GET"])
```
Comment (alan-wu) [inline app/main.py]: One last thing I noticed, I would prefer specifying the parameters after the route, 
i.e `/download/<program>/<project>/<uuid>/<format>`
Thanks

Comment (ddjnw1yu) [reply app/main.py]: Yes, I have modified it.

**Review by alan-wu — CHANGES_REQUESTED** [review]
Thanks for the changes, can you check with pytest. please? It is failing for me at this moment.

**Review by alan-wu — APPROVED** [review]
Thanks. Looks good to me.

**Pull request merged**